[testbed] Fix loadgenerator race condition

lahsivjar opened the pull request:
**Description:** <Describe what has changed.>
Fixes a race condition in testbed's default load generator (`ProviderSender`). The fix changes the behavior of how logging works for the `ProviderSender`. Before this PR, the code will try to log errors if the previous error is not the same and it would try to do this across goroutines. Now, each error will be logged if the previous error is not the same for each goroutine.

Alternatively, we can also build a logger using bloom filter to try to log each error once though I am not sure if that would be required. This PR offers a quick fix while keeping the behavior reasonably close to the current behavior.

Closes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32326

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32326

**Testing:** Follow the steps in the tracking issue.

**Documentation:** Not required.

songy23 reviewed (commented):
https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/8663186275/job/23756993964?pr=32351#step:8:189

Please fix linter

lahsivjar committed:
[testbed] Fix LoadGenerator race condition

ProviderSender based load generator had a race condition when used with
Parallel option configured to a value greater than 1. The commit fixes
the race condition by changing the behavior of how logging works w.r.t
previously observed errors.

lahsivjar commented:
@songy23 Thanks for the review. The previous fix was not totally correct so I have updated the fix to propagate the errors from `generate{Traces,Metrics,Logs}` to the `generate` func and handle previous error logging there. Also, the lint should be fixed too.

crobert-1 commented on testbed/testbed/load_generator.go:
```diff
@@ -210,22 +208,41 @@ func (ps *ProviderSender) generate() {
+						b, err = ps.generateLog()
 					default:
-						log.Printf("Invalid type of ProviderSender sender")
+						log.Printf("Invalid type of ProviderSender sender, quiting load generation")
```
```suggestion
						log.Printf("Invalid type of ProviderSender sender, quitting load generation")
```

crobert-1 commented on testbed/testbed/load_generator.go:
```diff
@@ -210,22 +208,41 @@ func (ps *ProviderSender) generate() {
+			var (
+				prevErr error
+				err     error
+				b       bool
```
Can we use a more descriptive name for the bool here? This coincides with another question, I'm not understanding exactly what the `bool` is supposed to represent that's being returned by the `generate...` methods. Can you explain what that refers to? I thought it was success at first, but that didn't seem quite right from usages.

lahsivjar replied:
Good point! Honestly, I am also unsure what the `bool` refers to. I wanted to move the decision of logging one level up and that meant moving all the parameters involved in the decision up one level. The `bool` comes from `DataProvider` ([ref](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/2daff18c2e3d9a1b63df72fd5c26f8c83a2d0683/testbed/testbed/data_providers.go#L28-L33)), however, the documentation is unclear on what it means.

Looking into the implementation ([ref](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/2daff18c2e3d9a1b63df72fd5c26f8c83a2d0683/testbed/testbed/data_providers.go#L219)) it seems that a `true` value refers to an empty(?) signal (Metric, Trace, or, Log). Based on the code in the load generator, empty signals are not processed further and thus don't impact `prevErr`. An option to simplify this would be to just clear the previous error even when the `bool` is `true`. It would mean that we would be logging with a bit more frequency in some cases, WDYT?

crobert-1 replied:
Thanks for sharing more information here. I came away with a similar understanding after looking a bit more into it, it looks like the bool _might_ be for error. So if an error occurs it'll be true, and false otherwise. (The only time it's true is as you shared, in the `goldenDataProvider`, and it looks like they're error cases, such as for `GenerateLogs` where it's not implemented.) It's almost like it's using a bool instead of just returning an error.

crobert-1 replied:
From our current understanding then, can we just remove `b`, and all references? Then update all of the `generate...()` methods to only return an error (instead of a bool and error)?

lahsivjar replied:
Updated to remove `b`. Also, a couple of points to highlight the changes that this will lead to:

1. Previously the `prevErr` would not reset when the `b` would be `true` (which as per our current understanding refers to an error in the provider). This means that now we would log a bit more if the error in provider resets the `prevErr` and then again a same error is observed. Practically, I am not sure if this would be a concern based on the current codebase.
2. I am not returning an error if `b` is true from the `generate{Log, Metrics, Trace}`. The main reason for this is that the current codebase doesn't seem to treat this case as an error. I don't have a strong opinion on this and would be happy to address this if you see this as incorrect.

crobert-1 commented on testbed/testbed/load_generator.go:
```diff
@@ -210,22 +208,41 @@ func (ps *ProviderSender) generate() {
 					default:
-						log.Printf("Invalid type of ProviderSender sender")
+						log.Printf("Invalid type of ProviderSender sender, quiting load generation")
+						continue
```
Why `continue` here instead of `return`? It seems to contradict the log message, and results in just coming back to the same case in the `select` from what I can tell, resulting in log spam without reason.

I'm not sure how returning here would impact test runs, but hopefully the failure reason would be obvious enough with the log message.

lahsivjar replied:
> and results in just coming back to the same case in the select from what I can tell, resulting in log spam without reason

I agree. The only reason I went with the `continue` was to keep the same behavior as before. I think we should just error if the signal type is unsupported in the [constructor itself](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/2daff18c2e3d9a1b63df72fd5c26f8c83a2d0683/testbed/testbed/load_generator.go#L98), WDYT?

crobert-1 replied:
I'd prefer to error sooner rather than later since this is for testing. No reason to waste time if it's going to fail. 👍

lahsivjar replied:
Updated the logic to error in the constructor if the sender type is not supported.

crobert-1 commented on testbed/testbed/load_generator.go:
```diff
@@ -210,22 +208,41 @@ func (ps *ProviderSender) generate() {
 			defer t.Stop()
+
+			var (
+				prevErr error
```
Is `prevErr` supposed to hold all previous errors, just the last error, or something else? I'm trying to better understand the point here before making more suggestions.

lahsivjar replied:
As per my understanding of the code,

1. Before this PR, the `prevErr` was supposed to hold the last error that occurred during load generation in any of the goroutines -- not sure if this behavior was intended. Also, this is what introduced the race condition that the PR is trying to solve.
2. After this PR, the `prevErr` holds the last error for each goroutine generating the load.

Alternatively, we can get rid of the `prevErr` completely and build a `onceLogger` that can use some probabilistic data structure (like bloom filter) to track what has happened before and only log new errors.

crobert-1 reviewed (approved):
Looks good to me, just one minor suggestion. Thanks for all of your work here, and helping me understand what's going on better! Really appreciate it!

lahsivjar committed:
Simply prevErr handling

lahsivjar committed:
Merge branch 'main' into fix-load-gen

lahsivjar commented on testbed/testbed/load_generator.go:
```diff
@@ -210,22 +212,26 @@ func (ps *ProviderSender) generate() {
+				if prevErr == nil || err.Error() != prevErr.Error() {
+					prevErr = err
+					log.Printf("%v", err)
+				}
```
Good point, thanks for raising it. I got tunnel-visioned with the previous approach to not see it 🤦 . I have updated the code now, let me know if it looks good to you.

crobert-1 replied:
Yep, looks good!

lahsivjar committed:
Merge branch 'main' into fix-load-gen

lahsivjar commented:
@songy23 @crobert-1 Is there anything blocking here to get this merged?

lahsivjar committed:
Merge branch 'main' into fix-load-gen

lahsivjar committed:
Merge branch 'main' into fix-load-gen

lahsivjar commented:
@atoulme Friendly ping, will you be able to review this whenever you have time and merge the PR if things look okay?

lahsivjar committed:
Merge branch 'main' into fix-load-gen

lahsivjar committed:
Merge branch 'main' into fix-load-gen

lahsivjar committed:
Merge branch 'main' into fix-load-gen

lahsivjar committed:
Merge branch 'main' into fix-load-gen

lahsivjar committed:
Merge branch 'main' into fix-load-gen
